### PR TITLE
Renaming and helpers

### PR DESCRIPTION
**Pull request by j042**
# Overview
- Covers #31
- add helpers directory
- use TrackedArray, adapted from [trimesh](https://github.com/mikedh/trimesh), thank you!
- add CI: checks format and runs test
- add simple test to check if all the functions without args are callable (TODO: extend!)
- formatted

### CODE REVIEW

File: gustaf/helpers/data.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+        return inner
+
+
+Unique2DFloats = namedtuple(
+    "Unique2DFloats",
+    [
+        "values",
+        "ids",
+        "inverse",
+        "union"
```
Comment (j042) [inline gustaf/helpers/data.py:441]:
```suggestion
        "intersection"
```

File: gustaf/vertices.py [hunk 2]
```diff
@@ -44,316 +40,142 @@ def __init__(
...
-
-    def get_vertices_overlapping(
-            self,
-            tolerance=settings.TOLERANCE,
-            referenced_only=True,
-            return_referenced=False,
-            workers=1,
-    ):
+        unique_vertices_info: Unique2DFloats
+          namedtuple with `values`, `ids`, `inverse`, `union`.
```
Comment (j042) [inline gustaf/vertices.py:159]:
```suggestion
          namedtuple with `values`, `ids`, `inverse`, `intersection`.
```

File: gustaf/vertices.py [hunk 3]
```diff
@@ -44,316 +40,142 @@ def __init__(
...
-        --------
-        self.vertices_overlapping: (len(self.vertices)) np.ndarray
-          list
-        """
-        last_item_is_ref = self.get_vertices_unique(
-            tolerance=tolerance,
-            referenced_only=referenced_only,
-            return_referenced=return_referenced,
-            workers=workers,
+        values, ids, inverse, union = utils.arr.close_rows(
```
Comment (j042) [inline gustaf/vertices.py:165]:
```suggestion
        values, ids, inverse, intersection = utils.arr.close_rows(
```

File: gustaf/vertices.py [hunk 4]
```diff
@@ -44,316 +40,142 @@ def __init__(
...
-        if return_referenced:
-            return self.vertices_overlapping, last_item_is_ref[-1]
-
-        else:
-            return self.vertices_overlapping
+        return helpers.data.Unique2DFloats(
+            values,
+            ids,
+            inverse,
+            union,
```
Comment (j042) [inline gustaf/vertices.py:174]: intersection

Comment (j042) [inline gustaf/vertices.py:174]:
```suggestion
            intersection,
```

File: gustaf/helpers/data.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+Note that for float types, "close enough" might be a better name than unique.
+This way, all tracked arrays, as long as they are 2D, have a dot separated
+syntax to acces unique info. For example, `mesh.unique_vertices.ids`.
+
+Attributes
+-----------
+values: (n, d) np.ndarray
+ids: (n) np.ndarray
+inverse: (m) np.ndarray
+union: (m) list of list
```
Comment (j042) [inline gustaf/helpers/data.py:455]:
```suggestion
intersection: (m) list of list
```

File: gustaf/vertices.py [hunk 6]
```diff
@@ -18,14 +19,9 @@ class Vertices(GustafBase):
...
-        "vertices_unique_inverse",
-        "vertices_overlapping",
-        "bounds",
-        "centers",
+        "_vertices",
+        "_const_vertices",
+        "_computed",
         "vis_dict",
         "vertexdata",
     ]
```
[the remark this answers is not in the document]
Comment (j042) [reply gustaf/vertices.py]: nice catch

File: gustaf/helpers/data.py [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+    """Taken from nice implementations of `trimesh` (see LICENSE.txt).
+    `https://github.com/mikedh/trimesh/blob/main/trimesh/caching.py`. Minor
+    adaption, since we don't have hashing functionalities.
+
+    All the inplace functions will set modified flag and if some operations
+    has potential to cause un-trackable behavior, writeable flags will be set
+    to False.
+
+    Note, if you really really want, it is possible to change the tracked
+    array with out setting modified flag.
```
Comment (jzwar) [inline gustaf/helpers/data.py:23]:
```suggestion
    array without setting modified flag.
```

File: gustaf/helpers/data.py [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+
+    def __ilshift__(self, *args, **kwargs):
+        self._set_modified()
+        return super(self.__class__, self).__ilshift__(*args, **kwargs)
+
+    def __irshift__(self, *args, **kwargs):
+        self._set_modified()
+        return super(self.__class__, self).__irshift__(*args, **kwargs)
+
+    def __iand__(self, *args, **kwargs):
```
Comment (jzwar) [inline gustaf/helpers/data.py:111]: I have to admit, this is far beyond my python expertise and I would have to google every single line in order to understand this. So I take this as is

**Review by jzwar — COMMENTED** [review]
The perfect Viennese word to describe this PR is probably "Oida". I wish you had put it in a few digestible chunks.

The style changes are nice. I would love to understand the tracked array, but for my taste it would need a bit more extensive documentation. I only had a brief look at some of the mesh-related parts of the code, but I am lacking the overview to be able to give constructive feedback here. 

Still - Overall, great work!

**Pull request merged**